### ISSUE TITLE
Change pattern for background media overlay

### ISSUE THREAD

**Issue by barklund** [GoogleForCreators/web-stories-wp#1207]
## Feature Description

It is currently possible to add a (semitransparent) overlay to a background image or video, but you cannot control this overlay in terms of color, opacity, color stops or gradient direction.

This will be partially solved by #403 allowing users to directly manipulate gradients on the canvas using handles (will allow manipulating size, position and rotation), but color changes has to be done using a color picker.

This feature is about adding a color picker for the background overlay, when an overlay is selected. It will be possible to change the type of overlay (solid, linear, radial) from inside the color picker, and the overlay display must react to this.

With no background overlay, no color picker is shown:
![Screen Shot 2020-04-14 at 2 56 49 PM](https://user-images.githubusercontent.com/637548/79263059-3c309b00-7e60-11ea-8604-49b5c5f8b020.png)

With background overlay set to something other than "none", color picker is shown:
![Screen Shot 2020-04-14 at 2 56 45 PM](https://user-images.githubusercontent.com/637548/79263053-3aff6e00-7e60-11ea-93b7-b9c28a6a5b8a.png)

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Implementation Brief

* Change `backgroundOverlay` on a page to be of type `Pattern` rather than the current Overlay enum.
* When clicking either of the buttons to select an overlay type, select a preset pattern similar to what is currently rendered for the overlay types.
* When a background overlay is set, add a color input component, that will open a color picker to update the color.
* Make sure to reflect the current pattern type in the select button in the overlay panel (might need to add a conic overlay type **or** configure color picker to not have conic gradient type for this specific value)
* The tricky thing here is when the user changes overlay type by pressing the buttons directly in the overlay panel. I suggest the following logic:
* * when switching from gradient to solid, use first color stop as solid color and gradient alpha as solid color alpha component (multiplied with any alpha already present on first color).
* * when switching from gradient to gradient, reuse stops,
* * when switching from solid to gradient, use solid color at alpha=0 as pos=0 and solid color at alpha=1 pos=1 color for gradient, and use solid color alpha component as gradient alpha.

### Migration
1. Before applying this PR, create a new story with 3 pages each with a background image, one with solid overlay, one with linear overlay and one with radial overlay
1. Migrate to this PR
1. See that all three pages still have the correct overlays
1. Do notice, that the radial overlay has changed slightly to a smoother and more round vignette.

### Edit overlay
1. Add a non-none background overlay to an image
1. Observe that you now can fully edit the color (and/or gradient) of the background overlay

Please report any errors that occur when playing around with overlays switching between types, etc. Note that the rules for switching between types are a bit complex, but very specifically designed to be exactly the way they are now (barring bugs).

**Comment by pbakaus** [comment]
Yep, description LGTM. Removing myself from the ticket for now, I think this can be implemented.

**Comment by jauyong** [comment]
@pbakaus per the above, I'm marking this for sprint 27 and adding the 1.0 Beta Milestone.

**Comment by barklund** [comment]
@pbakaus do we want to persist this overlay color between background elements and associate the information with the page overall? If I e.g. add a background element, add an overlay to it, change the color to blue and then replace the background element with another image, do we want to keep information on the new background image including a) that it has an overlay at all and/or b) the exact overlay including type and color (and color stops if gradient)?

Or do we simply associate the background overlay information directly with the element in question, so that if the element is removed (or replaced or unset as background), all overlay information is lost.

**Comment by barklund** [comment]
This issue might clash with potential changes upcoming in #1191, so I consider doing this after #1191 is merged.

**Comment by pbakaus** [comment]
> @pbakaus do we want to persist this overlay color between background elements and associate the information with the page overall? If I e.g. add a background element, add an overlay to it, change the color to blue and then replace the background element with another image, do we want to keep information on the new background image including a) that it has an overlay at all and/or b) the exact overlay including type and color (and color stops if gradient)?
> 
> Or do we simply associate the background overlay information directly with the element in question, so that if the element is removed (or replaced or unset as background), all overlay information is lost.

It should be associated with the page, not background media, as I think it would be fairly annoying to drag another image into the background after you carefully designed a nice overlay gradient, only for it to be wiped away then.

**Comment by csossi** [comment]
These QA steps failed:
"Before applying this PR, create a new story with 3 pages each with a background image, one with solid overlay, one with linear overlay and one with radial overlay
Migrate to this PR
See that all three pages still have the correct overlays"

Created the overlays as describe here in Master, but they disappeared when I switched to PR#1925

**Comment by csossi** [comment]
Unable to edit/select gradient colour if I select Linear or Radial

(here, I'm trying to select a red linear gradient)
![image.png](https://images.zenhubusercontent.com/5ea085e4704e72c0a99c6997/4fb2e6ee-7b58-421f-bf09-854c653b3d6e)

**Comment by barklund** [comment]
> Unable to edit/select gradient colour if I select Linear or Radial

You are changing the color of the first stop - it has 0% opacity. It will not actually show anything at 0% opacity:
![opacity](https://user-images.githubusercontent.com/637548/84526112-bc318080-acaa-11ea-9da2-d9f946600be1.png)

If you either:

1. Switch to the second stop
1. Change the opacity from 0%

You will see the color change appear on the resulting gradient.

That being said, it is a little weird, that clicking the color picker and changing the color doesn't immediately impact the gradient (as focus is on an "inactive" color stop), but that's how it works. Let's create that as a follow-up ticket for later.

**Comment by csossi** [comment]
Verified in QA

**Comment by samitron7** [comment]
Verified in UAT
